NIFI-1257 and 1259

alopresto opened the pull request:
NIFI-1257
NIFI-1259

Added the mechanics for additional strong `Key Derivation Functions` (KDF) including `PBKDF2`, `Bcrypt`, and `Scrypt` as well as `keyed ciphers` (raw key; no KDF). Integrated with `EncryptContent` processor and added new property and property descriptors. Dependent properties not available until NIFI-1121 is implemented. 

Explanation of KDFs available here: [Apache NiFi Wiki](https://cwiki.apache.org/confluence/display/NIFI/Key+Derivation+Function+Explanations)

alopresto commented:
The Travis CI failure is due to my use of `java.util.Base64` which was introduced in Java 8. Working on switching out to `org.apache.commons.codec.binary.Base64` but this does not provide an option to encode without padding unless using URL-safe character set (differing in "+" and "/" -> "-" and "_"). This causes compatibility issues in the Scrypt and Bcrypt salt formats.

markap14 commented on nifi-commons/nifi-security-utils/src/main/java/org/apache/nifi/security/util/KeyDerivationFunction.java:
```diff
@@ -44,6 +47,10 @@ public String getDescription() {
     }
 
+    public boolean isStrongKDF() {
+        return (name.equalsIgnoreCase(BCRYPT.name) || name.equals(SCRYPT.name) || name.equals(PBKDF2.name));
```
is there a reason for checking equalsIgnoreCase for BCRYPT but just equals for SCRIPT, PBKDF2?

alopresto replied:
Nope, will fix.

markap14 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/EncryptContent.java:
```diff
@@ -68,70 +73,78 @@
+            .build();
+    public static final PropertyDescriptor RAW_KEY_HEX = new PropertyDescriptor.Builder()
+            .name("raw-key-hex")
+            .displayName("Raw key (hexadecimal)")
```
Would recommend using Title Case for consistency with other processor properties

alopresto replied:
I copied the casing and style for `name` from `private-keyring-passphrase` which was an existing property descriptor. Or were you talking about the `displayName` attribute? I will fix that.

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/pom.xml:
```diff
@@ -124,4 +124,92 @@
+    <profiles>
+        <profile>
+            <!-- Custom profile for Groovy tests -->
+            <id>groovyUnitTest</id>
```
Do we need this profile with it already included in the root pom? Or if there something different about this one? Nit pick, existing profile id's naming convention ... groovy-unit-test.

alopresto replied:
No, this is an artifact because I was using Groovy tests internally before [NIFI-1365] was accepted. Will remove.

markap14 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/crypto/CipherUtility.java:
```diff
@@ -0,0 +1,271 @@
+    }
+
+    private static int parseActualKeyLengthFromAlgorithm(final String algorithm) {
+        Pattern pattern = Pattern.compile("([\\d]+)BIT");
```
Would recommend defining the Pattern as a 'private static final' member variable, as it is thread-safe and doesn't change

alopresto replied:
Will do.

markap14 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/crypto/CipherUtility.java:
```diff
@@ -0,0 +1,271 @@
+                }
+            }
+
+            try {
```
Can remove this try/catch, as the handling of Exception is the same as the outer try/catch

alopresto replied:
Good catch.

markap14 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/crypto/CipherUtility.java:
```diff
@@ -0,0 +1,271 @@
+        in.mark(limit);
+
+        // The first n bytes of the input stream contain the value up to the custom delimiter
+        if (in.available() < minimum) {
```
I think this logic may be incorrect - in.available() could return 0 or 1 or any number of values even if the stream has a huge amount of data available to be read, if reading it could result in blocking for I/O to occur...

alopresto replied:
After discussion with Mark, removed this check as (the intention) is redundant given the `StreamUtils.copyExclusive()` call also uses a limit.

apiri commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/crypto/OpenSSLPKCS5CipherProvider.java:
```diff
@@ -0,0 +1,211 @@
+    public Cipher getCipher(EncryptionMethod encryptionMethod, String password, byte[] salt, int keyLength, boolean encryptMode) throws Exception {
+        try {
+            return getInitializedCipher(encryptionMethod, password, salt, encryptMode);
+        } catch (IllegalArgumentException e) {
```
Is there a reason why this IllegalArgumentException is caught in lieu of catching/wrapping this with the associated ProcessException below?

alopresto replied:
During unit testing I was treating these differently. I guess in a production application the validation is not done any earlier, so it may not provide any value. I just figured that with crypto stuff, so much of it is opaque to a user that if I could separate out "you need to provide a good value for _X_" vs. "_magic_ went wrong" exceptions, that would be helpful.

alopresto commented:
The current state of the code passes on Java 7 and 8 with unlimited strength cryptographic jurisdiction policies installed (Travis includes these). I am resolving some tests which would fail on a vanilla JVM without these policies installed and will push these fixes. 

I also discovered some inconsistencies in the `BouncyCastle` implementation of certain PBE ciphers that can exceed the key strength limit even when using the correct key length if the underlying password length exceeds an undocumented inner limit. This was originally recorded in [NIFI-1255](https://issues.apache.org/jira/browse/NIFI-1255) but I have added a custom validator temporarily and added further documentation to that ticket.

apiri commented:
Reviewing latest.  Issue with build but only for one JDK, contrib and such look fine.

apiri commented:
@alopresto I'm running into issues with executing the built assembly without the JCE USJ files installed:

> 2016-02-04 10:52:14,023 ERROR [main] org.apache.nifi.NiFi Failure to launch NiFi due to java.util.ServiceConfigurationError: org.apache.nifi.controller.ControllerService: Provider org.apache.nifi.ssl.StandardSSLContextService could not be instantiated
java.util.ServiceConfigurationError: org.apache.nifi.controller.ControllerService: Provider org.apache.nifi.ssl.StandardSSLContextService could not be instantiated
	at java.util.ServiceLoader.fail(ServiceLoader.java:232) ~[na:1.8.0_60]
	at java.util.ServiceLoader.access$100(ServiceLoader.java:185) ~[na:1.8.0_60]
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:384) ~[na:1.8.0_60]
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404) ~[na:1.8.0_60]
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480) ~[na:1.8.0_60]
	at org.apache.nifi.nar.ExtensionManager.loadExtensions(ExtensionManager.java:107) ~[nifi-nar-utils-0.4.2-SNAPSHOT.jar:0.4.2-SNAPSHOT]
	at org.apache.nifi.nar.ExtensionManager.discoverExtensions(ExtensionManager.java:88) ~[nifi-nar-utils-0.4.2-SNAPSHOT.jar:0.4.2-SNAPSHOT]
	at org.apache.nifi.NiFi.<init>(NiFi.java:120) ~[nifi-runtime-0.4.2-SNAPSHOT.jar:0.4.2-SNAPSHOT]
	at org.apache.nifi.NiFi.main(NiFi.java:227) ~[nifi-runtime-0.4.2-SNAPSHOT.jar:0.4.2-SNAPSHOT]
Caused by: java.lang.ExceptionInInitializerError: null
	at javax.crypto.JceSecurityManager.<clinit>(JceSecurityManager.java:65) ~[na:1.8.0_60]
	at javax.crypto.Cipher.getConfiguredPermission(Cipher.java:2587) ~[na:1.8.0_60]
	at javax.crypto.Cipher.getMaxAllowedKeyLength(Cipher.java:2611) ~[na:1.8.0_60]
	at sun.security.ssl.CipherSuite$BulkCipher.isAvailable(CipherSuite.java:548) ~[na:1.8.0_60]
	at sun.security.ssl.CipherSuite$BulkCipher.isAvailable(CipherSuite.java:527) ~[na:1.8.0_60]
	at sun.security.ssl.CipherSuite.isAvailable(CipherSuite.java:194) ~[na:1.8.0_60]
	at sun.security.ssl.SSLContextImpl.getApplicableCipherSuiteList(SSLContextImpl.java:346) ~[na:1.8.0_60]
	at sun.security.ssl.SSLContextImpl.getDefaultCipherSuiteList(SSLContextImpl.java:297) ~[na:1.8.0_60]
	at sun.security.ssl.SSLEngineImpl.init(SSLEngineImpl.java:402) ~[na:1.8.0_60]
	at sun.security.ssl.SSLEngineImpl.<init>(SSLEngineImpl.java:349) ~[na:1.8.0_60]
	at sun.security.ssl.SSLContextImpl.engineCreateSSLEngine(SSLContextImpl.java:201) ~[na:1.8.0_60]
	at javax.net.ssl.SSLContext.createSSLEngine(SSLContext.java:329) ~[na:1.8.0_60]
	at org.apache.nifi.ssl.StandardSSLContextService.buildAlgorithmAllowableValues(StandardSSLContextService.java:424) ~[nifi-ssl-context-service-0.4.2-SNAPSHOT.jar:0.4.2-SNAPSHOT]
	at org.apache.nifi.ssl.StandardSSLContextService.<clinit>(StandardSSLContextService.java:103) ~[nifi-ssl-context-service-0.4.2-SNAPSHOT.jar:0.4.2-SNAPSHOT]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_60]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:1.8.0_60]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.8.0_60]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422) ~[na:1.8.0_60]
	at java.lang.Class.newInstance(Class.java:442) ~[na:1.8.0_60]
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:380) ~[na:1.8.0_60]
	... 6 common frames omitted
Caused by: java.lang.SecurityException: Can not initialize cryptographic mechanism
	at javax.crypto.JceSecurity.<clinit>(JceSecurity.java:89) ~[na:1.8.0_60]
	... 26 common frames omitted
Caused by: java.lang.SecurityException: Cannot locate policy or framework files!
	at javax.crypto.JceSecurity.setupJurisdictionPolicies(JceSecurity.java:256) ~[na:1.8.0_60]
	at javax.crypto.JceSecurity.access$000(JceSecurity.java:48) ~[na:1.8.0_60]
	at javax.crypto.JceSecurity$1.run(JceSecurity.java:81) ~[na:1.8.0_60]
	at java.security.AccessController.doPrivileged(Native Method) ~[na:1.8.0_60]
	at javax.crypto.JceSecurity.<clinit>(JceSecurity.java:78) ~[na:1.8.0_60]
	... 26 common frames omitted

With them installed, NiFi starts and works as anticipated.  Will dig in a bit more to see what changed to cause the issue within
	at org.apache.nifi.ssl.StandardSSLContextService.buildAlgorithmAllowableValues(StandardSSLContextService.java:424)

alopresto commented:
Worked with Aldrin and determined that the `brew` method of installing/uninstalling JCE USJ files is not correct, as "uninstalling" removes the unlimited jurisdiction policy files but does not replace them with the originals.

apiri commented on nifi-commons/nifi-security-utils/src/main/java/org/apache/nifi/security/util/KeyDerivationFunction.java:
```diff
@@ -25,8 +25,11 @@
-    OPENSSL_EVP_BYTES_TO_KEY("OpenSSL EVP_BytesToKey", "Single iteration MD5 compatible with PKCS#5 v1.5");
-    // TODO: Implement bcrypt, scrypt, and PBKDF2
+    OPENSSL_EVP_BYTES_TO_KEY("OpenSSL EVP_BytesToKey", "Single iteration MD5 compatible with PKCS#5 v1.5"),
+    BCRYPT("Bcrypt", "Bcrypt with configurable work factor: see https://cwiki.apache.org/confluence/display/NIFI/Key+Derivation+Function+Explanations"),
```
The resources listed on the wiki are super helpful. I think it would make sense for these to additionally get rolled into the administrator's guide so we can have a one stop shop for all of it.

alopresto replied:
Working on that now. Converting various markup languages.

mattyb149 commented on nifi-commons/nifi-security-utils/src/main/java/org/apache/nifi/security/util/EncryptionMethod.java:
```diff
@@ -26,37 +26,43 @@
+    private final boolean compatibleWithStrongKDFs;
 
-    EncryptionMethod(String algorithm, String provider, boolean unlimitedStrength) {
+    EncryptionMethod(String algorithm, String provider, boolean unlimitedStrength, boolean compatibleWithStrongKDFs) {
```
Can this be a straight override or do we need to keep the original method and deprecate it?

alopresto replied:
Because this is the enum constructor, it's never used anywhere else.

mattyb149 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/groovy/org/apache/nifi/processors/standard/util/crypto/CipherUtilityGroovyTest.groovy:
```diff
@@ -0,0 +1,251 @@
+        }
+
+        // Extra hard-coded checks
+        ["PBEWITHSHA256AND256BITAES-CBC-BC": 192].each { String algorithm, int invalidKeyLength ->
```
Let's add extra entries here (or do away with the map)

alopresto replied:
Done.

apiri commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/resources/logback-test.xml:
```diff
@@ -57,11 +60,12 @@
 
-    <root level="INFO">
+    <!--<root level="${overrideLog:-DEBUG}">-->
+    <root level="DEBUG">
```
Did you intend to revert this back to the override log variable and default to INFO?

alopresto replied:
I never actually got this to work. Reverting.
